olp-cpp-sdk-core/src/http/winhttp/NetworkWinHttp.cpp

```diff
@@ -0,0 +1,1006 @@
+                          &out.front(), out.size());
+
+  // Should not happen as 1st call have succeeded.
+  return conversion_result != 0;
+}
+
+std::wstring ProxyString(const olp::http::NetworkProxySettings& proxy) {
+  using namespace olp::http;
+  std::wostringstream proxy_string_stream;
+
+  switch (proxy.GetType()) {
+    case NetworkProxySettings::Type::NONE:
+      proxy_string_stream << "http://";
+      break;
+    case NetworkProxySettings::Type::SOCKS4:
+      proxy_string_stream << "socks4://";
+      break;
+    case NetworkProxySettings::Type::SOCKS5:
+      proxy_string_stream << "socks5://";
+      break;
+    case NetworkProxySettings::Type::SOCKS4A:
+      proxy_string_stream << "socks4a://";
+      break;
+    case NetworkProxySettings::Type::SOCKS5_HOSTNAME:
+      proxy_string_stream << "socks5h://";
+      break;
+    default:
+      proxy_string_stream << "http://";
+  }
+
+  proxy_string_stream << std::wstring(proxy.GetUsername().begin(),
+                                      proxy.GetUsername().end());
+  proxy_string_stream << ':';
+  proxy_string_stream << proxy.GetPort();
+
+  return proxy_string_stream.str();
+}
+
+}  // namespace
+
+namespace olp {
+namespace http {
+
+#define LOGTAG "WinHttp"
+
+NetworkWinHttp::NetworkWinHttp()
+    : http_session_(NULL),
+      thread_(INVALID_HANDLE_VALUE),
+      event_(INVALID_HANDLE_VALUE) {
+  request_id_counter_.store(
+      static_cast<RequestId>(RequestIdConstants::RequestIdMin));
+
+  http_session_ = WinHttpOpen(L"OLP Http Client", WINHTTP_ACCESS_TYPE_NO_PROXY,
+                              WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS,
+                              WINHTTP_FLAG_ASYNC);
+
+  if (!http_session_) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpOpen failed " << GetLastError());
+    return;
+  }
+
+  WinHttpSetStatusCallback(
+      http_session_, (WINHTTP_STATUS_CALLBACK)&NetworkWinHttp::RequestCallback,
+      WINHTTP_CALLBACK_FLAG_ALL_COMPLETIONS | WINHTTP_CALLBACK_FLAG_HANDLES, 0);
+
+  event_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+  // Store the memory tracking state during initialization so that it can be
+  // used by the thread.
+  thread_ = CreateThread(NULL, 0, NetworkWinHttp::Run, this, 0, NULL);
+  SetThreadPriority(thread_, THREAD_PRIORITY_ABOVE_NORMAL);
+}
+
+NetworkWinHttp::~NetworkWinHttp() {
+  std::vector<std::shared_ptr<ResultData>> pendingResults;
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    std::vector<std::shared_ptr<RequestData>> requestsToCancel;
+    for (auto& request : http_requests_) {
+      requestsToCancel.push_back(request.second);
+    }
+    while (!requestsToCancel.empty()) {
+      std::shared_ptr<RequestData> request = requestsToCancel.front();
+      WinHttpCloseHandle(request->http_request);
+      request->http_request = NULL;
+      pendingResults.push_back(request->result_data);
+      requestsToCancel.erase(requestsToCancel.begin());
```

Comment:
Erasing vector at the beginning is costly (O(N)) operation. Better use`std::list` instead.

But I don't get why to need to mutate the `requestsToCancel` at all? We can just iterate over the container of requests via for loop. Or even better - jut to iterate just over original `http_requests_` without creating, WDYT?

Reply:
Right now it's a protection against new send requests from cancel callback. Iteration over http_requests_ could end in an endless loop. But this should be changed, during improvement.

Reply:
What about std::list / std::queue?

Anyway, if you want to address this issue during refactoring phase, please, add corresponding item into our improvements page, so we don't forget it.